From 6ffff751cdf6da3fa19f033655f710363bb133c0 Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Tue, 11 Aug 2026 14:03:57 -0300
Subject: [PATCH] pinctrl: airoha: align EN751221 PHY LED naming

Use the same phyN_led0 function naming and GPIO group model as the
EN7523 and AN7581 pinctrl data. The EN751221 exposes one routed LED
pad for each of its four internal PHY-facing switch ports. Keep the
additional switch port 4 output separate as gsw_port4_led0 rather than
inventing a fifth PHY.

Also route the PON I2C path when selecting the PON function, as required
by the EN7521 vendor SDK.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 .../pinctrl/econet,en751221-pinctrl.yaml      | 39 +++++++++-
 drivers/pinctrl/airoha/pinctrl-en751221.c     | 76 +++++++++++--------
 2 files changed, 79 insertions(+), 36 deletions(-)

diff --git a/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml b/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
index 0452b931e604..2685e9ac5aa6 100644
--- a/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
+++ b/Documentation/devicetree/bindings/pinctrl/econet,en751221-pinctrl.yaml
@@ -65,7 +65,8 @@ patternProperties:
         properties:
           function:
             enum: [sipo, tod_1pps, pcie_reset, spi, uart, pon, pcm,
-                   pcm_spi, phy_led, gpio]
+                   pcm_spi, gpio, phy1_led0, phy2_led0, phy3_led0,
+                   phy4_led0, gsw_port4_led0]
 
           groups: true
 
@@ -144,11 +145,43 @@ patternProperties:
           - if:
               properties:
                 function:
-                  const: phy_led
+                  const: phy1_led0
             then:
               properties:
                 groups:
-                  enum: [phy0_led, phy1_led, phy2_led, phy3_led, phy4_led]
+                  enum: [gpio10]
+          - if:
+              properties:
+                function:
+                  const: phy2_led0
+            then:
+              properties:
+                groups:
+                  enum: [gpio9]
+          - if:
+              properties:
+                function:
+                  const: phy3_led0
+            then:
+              properties:
+                groups:
+                  enum: [gpio8]
+          - if:
+              properties:
+                function:
+                  const: phy4_led0
+            then:
+              properties:
+                groups:
+                  enum: [gpio7]
+          - if:
+              properties:
+                function:
+                  const: gsw_port4_led0
+            then:
+              properties:
+                groups:
+                  enum: [gpio3]
           - if:
               properties:
                 function:
diff --git a/drivers/pinctrl/airoha/pinctrl-en751221.c b/drivers/pinctrl/airoha/pinctrl-en751221.c
index baf3fccc9473..c2b5cc85e2fd 100644
--- a/drivers/pinctrl/airoha/pinctrl-en751221.c
+++ b/drivers/pinctrl/airoha/pinctrl-en751221.c
@@ -25,24 +25,25 @@
 #define PCM_SPI_RESET_MODE_MASK		BIT(10)
 #define PCM_SPI_CS4_MODE_MASK			BIT(9)
 #define PCM_SPI_CS3_MODE_MASK			BIT(8)
-#define PHY4_LED_MODE_MASK			BIT(7)
-#define PHY3_LED_MODE_MASK			BIT(6)
-#define PHY2_LED_MODE_MASK			BIT(5)
-#define PHY1_LED_MODE_MASK			BIT(4)
-#define PHY0_LED_MODE_MASK			BIT(3)
+#define GSW_PORT4_LED0_MODE_MASK		BIT(7)
+#define PHY4_LED0_MODE_MASK			BIT(6)
+#define PHY3_LED0_MODE_MASK			BIT(5)
+#define PHY2_LED0_MODE_MASK			BIT(4)
+#define PHY1_LED0_MODE_MASK			BIT(3)
 #define PON_TOD_1PPS_MODE_MASK			BIT(2)
 #define GSW_TOD_1PPS_MODE_MASK			BIT(1)
+#define PON_I2C_MODE_MASK			BIT(0)
 
 #define GPIO3_MUX_MASK		(SPI_QUAD_MODE_MASK | UART2_MODE_MASK | \
 				 PCM_SPI_INT_MODE_MASK | PCM_SPI_CS3_MODE_MASK | \
-				 PHY4_LED_MODE_MASK)
+				 GSW_PORT4_LED0_MODE_MASK)
 #define GPIO4_6_MUX_MASK	(PCM2_MODE_MASK | PCM_SPI_MODE_MASK)
-#define GPIO7_MUX_MASK		(GPIO4_6_MUX_MASK | PHY3_LED_MODE_MASK)
-#define GPIO8_MUX_MASK		(SIPO_MODE_MASK | PHY2_LED_MODE_MASK)
+#define GPIO7_MUX_MASK		(GPIO4_6_MUX_MASK | PHY4_LED0_MODE_MASK)
+#define GPIO8_MUX_MASK		(SIPO_MODE_MASK | PHY3_LED0_MODE_MASK)
 #define GPIO9_MUX_MASK		(SIPO_RCLK_MODE_MASK | PCM_SPI_CS4_MODE_MASK | \
-				 PHY1_LED_MODE_MASK)
+				 PHY2_LED0_MODE_MASK)
 #define GPIO10_MUX_MASK		(SPI_QUAD_MODE_MASK | UART2_MODE_MASK | \
-				 PHY0_LED_MODE_MASK)
+				 PHY1_LED0_MODE_MASK)
 #define GPIO22_MUX_MASK		(DMT_TOD_1PPS_MODE_MASK | \
 				 PON_TOD_1PPS_MODE_MASK | \
 				 GSW_TOD_1PPS_MODE_MASK)
@@ -149,11 +150,6 @@ static const int pcm_spi_int_pins[] = { 3 };
 static const int pcm_spi_rst_pins[] = { 2 };
 static const int pcm_spi_cs3_pins[] = { 3 };
 static const int pcm_spi_cs4_pins[] = { 9 };
-static const int phy0_led_pins[] = { 10 };
-static const int phy1_led_pins[] = { 9 };
-static const int phy2_led_pins[] = { 8 };
-static const int phy3_led_pins[] = { 7 };
-static const int phy4_led_pins[] = { 3 };
 static const int gpio2_pins[] = { 2 };
 static const int gpio3_pins[] = { 3 };
 static const int gpio4_pins[] = { 4 };
@@ -195,11 +191,6 @@ static const struct pingroup pinctrl_groups[] = {
 	PINCTRL_PIN_GROUP("pcm_spi_rst", pcm_spi_rst),
 	PINCTRL_PIN_GROUP("pcm_spi_cs3", pcm_spi_cs3),
 	PINCTRL_PIN_GROUP("pcm_spi_cs4", pcm_spi_cs4),
-	PINCTRL_PIN_GROUP("phy0_led", phy0_led),
-	PINCTRL_PIN_GROUP("phy1_led", phy1_led),
-	PINCTRL_PIN_GROUP("phy2_led", phy2_led),
-	PINCTRL_PIN_GROUP("phy3_led", phy3_led),
-	PINCTRL_PIN_GROUP("phy4_led", phy4_led),
 	PINCTRL_PIN_GROUP("gpio2", gpio2),
 	PINCTRL_PIN_GROUP("gpio3", gpio3),
 	PINCTRL_PIN_GROUP("gpio4", gpio4),
@@ -242,9 +233,11 @@ static const char *const pcm_spi_groups[] = {
 	"pcm_spi_cs4"
 };
 
-static const char *const phy_led_groups[] = {
-	"phy0_led", "phy1_led", "phy2_led", "phy3_led", "phy4_led"
-};
+static const char *const phy1_led0_groups[] = { "gpio10" };
+static const char *const phy2_led0_groups[] = { "gpio9" };
+static const char *const phy3_led0_groups[] = { "gpio8" };
+static const char *const phy4_led0_groups[] = { "gpio7" };
+static const char *const gsw_port4_led0_groups[] = { "gpio3" };
 
 static const char *const gpio_groups[] = {
 	"gpio2", "gpio3", "gpio4", "gpio5", "gpio6", "gpio7",
@@ -287,17 +280,18 @@ static const struct airoha_pinctrl_func_group uart_func_group[] = {
 };
 
 static const struct airoha_pinctrl_func_group pon_func_group[] = {
-	ECONET_MUX_GROUP("pon", PON_MODE_MASK, PON_MODE_MASK),
+	ECONET_MUX_GROUP("pon", PON_MODE_MASK | PON_I2C_MODE_MASK,
+			 PON_MODE_MASK | PON_I2C_MODE_MASK),
 };
 
 static const struct airoha_pinctrl_func_group pcm_func_group[] = {
 	ECONET_MUX_GROUP("pcm1", PCM1_MODE_MASK, PCM1_MODE_MASK),
-	ECONET_MUX_GROUP("pcm2", PCM2_MODE_MASK | PHY3_LED_MODE_MASK,
+	ECONET_MUX_GROUP("pcm2", PCM2_MODE_MASK | PHY4_LED0_MODE_MASK,
 			 PCM2_MODE_MASK),
 };
 
 static const struct airoha_pinctrl_func_group pcm_spi_func_group[] = {
-	ECONET_MUX_GROUP("pcm_spi", PCM_SPI_MODE_MASK | PHY3_LED_MODE_MASK,
+	ECONET_MUX_GROUP("pcm_spi", PCM_SPI_MODE_MASK | PHY4_LED0_MODE_MASK,
 			 PCM_SPI_MODE_MASK),
 	ECONET_MUX_GROUP("pcm_spi_int", GPIO3_MUX_MASK,
 			 PCM_SPI_INT_MODE_MASK),
@@ -309,12 +303,24 @@ static const struct airoha_pinctrl_func_group pcm_spi_func_group[] = {
 			 PCM_SPI_CS4_MODE_MASK),
 };
 
-static const struct airoha_pinctrl_func_group phy_led_func_group[] = {
-	ECONET_MUX_GROUP("phy0_led", GPIO10_MUX_MASK, PHY0_LED_MODE_MASK),
-	ECONET_MUX_GROUP("phy1_led", GPIO9_MUX_MASK, PHY1_LED_MODE_MASK),
-	ECONET_MUX_GROUP("phy2_led", GPIO8_MUX_MASK, PHY2_LED_MODE_MASK),
-	ECONET_MUX_GROUP("phy3_led", GPIO7_MUX_MASK, PHY3_LED_MODE_MASK),
-	ECONET_MUX_GROUP("phy4_led", GPIO3_MUX_MASK, PHY4_LED_MODE_MASK),
+static const struct airoha_pinctrl_func_group phy1_led0_func_group[] = {
+	ECONET_MUX_GROUP("gpio10", GPIO10_MUX_MASK, PHY1_LED0_MODE_MASK),
+};
+
+static const struct airoha_pinctrl_func_group phy2_led0_func_group[] = {
+	ECONET_MUX_GROUP("gpio9", GPIO9_MUX_MASK, PHY2_LED0_MODE_MASK),
+};
+
+static const struct airoha_pinctrl_func_group phy3_led0_func_group[] = {
+	ECONET_MUX_GROUP("gpio8", GPIO8_MUX_MASK, PHY3_LED0_MODE_MASK),
+};
+
+static const struct airoha_pinctrl_func_group phy4_led0_func_group[] = {
+	ECONET_MUX_GROUP("gpio7", GPIO7_MUX_MASK, PHY4_LED0_MODE_MASK),
+};
+
+static const struct airoha_pinctrl_func_group gsw_port4_led0_func_group[] = {
+	ECONET_MUX_GROUP("gpio3", GPIO3_MUX_MASK, GSW_PORT4_LED0_MODE_MASK),
 };
 
 static const struct airoha_pinctrl_func_group gpio_func_group[] = {
@@ -351,7 +357,11 @@ static const struct airoha_pinctrl_func pinctrl_funcs[] = {
 	PINCTRL_FUNC_DESC("pon", pon),
 	PINCTRL_FUNC_DESC("pcm", pcm),
 	PINCTRL_FUNC_DESC("pcm_spi", pcm_spi),
-	PINCTRL_FUNC_DESC("phy_led", phy_led),
+	PINCTRL_FUNC_DESC("phy1_led0", phy1_led0),
+	PINCTRL_FUNC_DESC("phy2_led0", phy2_led0),
+	PINCTRL_FUNC_DESC("phy3_led0", phy3_led0),
+	PINCTRL_FUNC_DESC("phy4_led0", phy4_led0),
+	PINCTRL_FUNC_DESC("gsw_port4_led0", gsw_port4_led0),
 	PINCTRL_FUNC_DESC("gpio", gpio),
 };
 
-- 
2.55.0

